### ISSUE TITLE
Modify-by-reference silently fails to .save() 

### ISSUE THREAD

**Issue by mwalmsley** [zooniverse/panoptes-python-client#211]
When modifying a PanoptesObject, the client tracks changes like so:

def __setattr__(self, name, value):
...
self.modified_attributes.add(name)
    
However, when modifying an attribute by reference rather than by __setattr__, e.g. workflow.configuration['hello'] = 'world' or workflow.configuration.update({hello: world}), self.modified_attributes remains empty. In consequence, self.save() has no effect on existing objects:

def save(self):
...
        if not self.id:
...
        else:
            if not self.modified_attributes:
                return  # this silently does nothing

Any __setattr__ call will work around this issue, e.g. workflow.configuration = workflow.configuration will trigger self.modified_attributes to update. However, this is not ideal.

The client could be extended to:
- print or log a warning, or even an exception, when .save() is called on unmodified objects and so no PUT is made
- monitor changes to mutable attributes, perhaps by wrapping dict and list, so that modifications-by-reference are noted

Finally, the user guide includes this line:
`subject.metadata['my_metadata'] = 'abcd'`
which may suffer from this issue.

@camallen  @adammcmaster

**Comment by adammcmaster** [comment]
`Subject` has code to handle that for metadata, by saving a copy in `self._original_metadata` and comparing them on save. We should generalise that in `PanoptesObject` so that it works for all dict attributes.

**Comment by camallen** [comment]
cc'ing @noraeisner - just ran into this with her as well.

**Comment by camallen** [comment]
was this issue closed by #222?

**Comment by lcjohnso** [comment]
> was this issue closed by #222?

No -- #222 allows saves of edited workflow configurations *only*, and continues to fail silently for other cases (e.g., retirement criteria).  A more general solution for other editable dict attributes is still needed.

Workaround for retirement criteria changes on workflow: add non-configuration to-be-updated items manually to modified_attributes before save. For example, for `retirement`:
```
workflow.retirement['criteria']='never_retire'
workflow.modified_attributes.add('retirement')
workflow.save()
```

**Comment by lcjohnso** [comment]
Closed by #251
